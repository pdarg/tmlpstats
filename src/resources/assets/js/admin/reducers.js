import { combineReducers } from 'redux'

import { quartersData } from './data'
import regionsReducer from './regions/reducers'

const lookupsReducer = combineReducers({
    quarters: quartersData.reducer()
})

const adminReducer = combineReducers({
    regions: regionsReducer,
    lookups: lookupsReducer
})

export default adminReducer
