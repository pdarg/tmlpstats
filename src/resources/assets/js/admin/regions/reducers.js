import { formReducer, modelReducer } from 'react-redux-form'
import { combineReducers } from 'redux'
import { regionsData, centersData, scoreboardLockData, extraData } from './data'

const regionsReducer = combineReducers({
    form: formReducer('admin.regions'),
    regions: regionsData.reducer(),
    centers: centersData.reducer(),
    scoreboardLock: scoreboardLockData.reducer(),
    extra: extraData.reducer(),
    quarterDates: modelReducer('admin.regions.quarterDates'),
})

export default regionsReducer
