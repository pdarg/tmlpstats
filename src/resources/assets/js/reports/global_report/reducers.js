import { reportData } from './data'

export default reportData.reducer()
