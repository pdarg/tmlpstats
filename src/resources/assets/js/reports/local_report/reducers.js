import { combineReducers } from 'redux'

import { reportData } from './data'

export default combineReducers({
    reportData: reportData.reducer(),
})
