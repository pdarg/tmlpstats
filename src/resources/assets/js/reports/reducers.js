import { combineReducers } from 'redux'

import globalReportReducer from './global_report/reducers'
import localReportReducer from './local_report/reducers'

function showTokenReducer(state=false, action) {
    if (action.type == 'reports/showToken') {
        return action.payload
    }
    return state
}

export default combineReducers({
    global_report: globalReportReducer,
    local_report: localReportReducer,
    showToken: showTokenReducer,
})
