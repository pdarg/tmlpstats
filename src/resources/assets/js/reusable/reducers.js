import { objectAssign } from './ponyfill'

const loadStates = {
    new: {state: 'new', loaded: false, available: true},
    loading: {state: 'loading', loaded: false, available: false},
    loaded: {state: 'loaded', loaded: true, available: true},
    failed: {state: 'failed', loaded: false, failed: true, available: false}
}

export function loadingMultiState(actionType) {
    return function(state = loadStates['new'], action){
        if (action.type == actionType) {
            // toUpperase check proves this is a string
            if (action.payload.toUpperCase) {
                return loadStates[action.payload]
            } else {
                return action.payload
            }
        }
        return state
    }
}

export class LoadingMultiState {
    static states = loadStates

    constructor(actionType) {
        this.actionType = actionType
    }

    actionCreator() {
        const actionType = this.actionType
        return (newState, extra) => {
            if (newState.error) {
                newState = objectAssign({}, loadStates.failed, newState)
            } else if (extra) {
                newState = objectAssign({}, loadStates[newState], extra)
            }
            return {type: actionType, payload: newState}
        }
    }

    reducer() {
        return loadingMultiState(this.actionType)
    }
}

const REPLACE_MESSAGES = 'messageManager/replace'
const REPLACE_ALL_MESSAGES = 'messageManager/reset'

export class MessageManager {
    constructor(namespace) {
        this.namespace = namespace
    }

    reducer() {
        return (state={}, action) => {
            if (action.payload && action.payload.namespace && action.payload.namespace == this.namespace) {
                switch (action.type) {
                case REPLACE_MESSAGES:
                    return objectAssign({}, state, action.payload.messages)
                case REPLACE_ALL_MESSAGES:
                    return action.payload.messages
                }
            }
            return state
        }
    }

    replaceAll(messages, namespace) {
        return {type: REPLACE_ALL_MESSAGES, payload: {messages, namespace: namespace || this.namespace}}
    }

    replaceMany(messageMap) {
        return {type: REPLACE_MESSAGES, payload: {messages: messageMap, namespace: this.namespace}}
    }

    replace(pk, messages) {
        return this.replaceMany({[pk]: messages})
    }

    reset() {
        return this.replaceAll({})
    }
}

const INLINE_BULK_DEFAULT = {ts: 0, changed: {}, working: null}

const IB_MARK = 'mark'
const IB_BEGIN_WORK = 'begin_work'
const IB_END_WORK = 'end_work'
const IB_CLEAR_WORK = 'clear_work'

/** Tracks bulk update work to some resource with a primary key. The idea is that we'd be live updating something. */
export class InlineBulkWork {
    constructor(prefix) {
        this.actionType = prefix
    }

    mark(pk) {
        return {type: this.actionType, payload:{job: IB_MARK, pk}}
    }

    beginWork() {
        return {type: this.actionType, payload:{job: IB_BEGIN_WORK}}
    }

    endWork() {
        return {type: this.actionType, payload:{job: IB_END_WORK}}
    }

    clearWork() {
        return {type: this.actionType, payload:{job: IB_CLEAR_WORK}}
    }

    reducer() {
        return (state=INLINE_BULK_DEFAULT, action) => {
            if (action.type == this.actionType) {
                const ts = state.ts + 1 // TS is an atomic incrementing timestamp, we can use this to see what happened 'after' other things.
                var { changed, working } = state
                const { job, pk } = action.payload
                switch (job) {
                case IB_MARK:
                    if (!changed[pk] || (working && working[pk])) {
                        changed = objectAssign({}, changed, {[pk]: ts})
                        return {ts, working, changed}
                    }
                    break
                case IB_BEGIN_WORK:
                    // Cool trick: set the currently working as the ones who have changed at this point, and we get the answer for free
                    return {ts, working: changed, changed}
                case IB_END_WORK:
                    // When work ends, we only want to clear changed items which have not changed yet again
                    changed = objectAssign({}, changed)
                    for (var k in working) {
                        if (changed[k] <= working[k]) {
                            delete changed[k]
                        }
                    }
                    return {ts, changed, working: null}
                }
            }
            return state
        }
    }
}
