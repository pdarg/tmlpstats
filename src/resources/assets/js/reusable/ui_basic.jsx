import React, { PropTypes } from 'react'

import { getErrMessage } from './ajax_utils'

/**
 * ModeSelectButtons renders side-by-side buttons used for mode selector, with appropriate active state.
 *
 * Typical props:
 *   - items: array of objects describing the item. Typically, the item must have props 'key' and 'label'
 *   - current: The key value of the currently selected item. string or number.
 *   - onClick: a callback which will be called with the selected key on item change.
 *
 * Optional props:
 *   - ariaGroupDesc: Used to provide an ARIA label to the button group which describes what it does.
 */
export class ModeSelectButtons extends React.Component {
    static defaultProps = {
        classes: 'btn btn-default',
        activeClasses: 'btn btn-default active',
        keyProp: 'key',
        getLabel: (v) => v.label
    }
    static propTypes = {
        items: PropTypes.oneOfType([
            PropTypes.object,
            PropTypes.arrayOf(PropTypes.object),
        ]).isRequired,
        current: PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number
        ]),
        ariaGroupDesc: PropTypes.string,
        onClick: PropTypes.func
    }

    render() {
        var buttons = []
        this.props.items.forEach((item) => {
            var key = item[this.props.keyProp]
            const classes = (key == this.props.current) ? this.props.activeClasses : this.props.classes
            var cb = () => this.props.onClick(key, item)
            buttons.push(
                <button key={key} type="button" className={classes} onClick={cb}>
                    {this.props.getLabel(item)}
                </button>
            )
        })

        return (
            <div className="btn-group" role="group" aria-label={this.props.ariaGroupDesc}>
                {buttons}
            </div>
        )
    }
}

/**
 * LoadStateFlip is a really simple experimental component for choosing what's shown based on load state.
 *
 * This should
 */
export class LoadStateFlip extends React.PureComponent {
    static propTypes = {
        loadState: PropTypes.shape({
            state: PropTypes.string.isRequired
        }).isRequired
    }
    render() {
        var loadState = this.props.loadState
        if (loadState.state == 'loading') {
            return <div><span className="glyphicon glyphicon-send"></span>TODO spinner here...</div>
        } else if (loadState.state == 'failed') {
            var error = 'FAILED '
            if (loadState.error) {
                error += getErrMessage(loadState.error)
            }
            return <div className="bg-danger">{error}</div>
        } else {
            return <div>{this.props.children}</div>
        }
    }
}

export class SubmitFlip extends React.PureComponent {
    static defaultProps = {
        buttonClasses: 'btn btn-primary',
        offset: 'col-md-offset-2 col-md-8',
        wrapGroup: true
    }
    static propTypes = {
        buttonClasses: PropTypes.string,
        loadState: PropTypes.object.isRequired,
        offset: PropTypes.string,
        wrapGroup: PropTypes.bool,
        children: PropTypes.node
    }

    render() {
        const { wrapGroup, offset, loadState, buttonClasses } = this.props

        const body = (
            <LoadStateFlip loadState={loadState}>
                <button className={buttonClasses} type="submit">{this.props.children}</button>
            </LoadStateFlip>
        )

        if (wrapGroup) {
            return (
                <div className="form-group">
                    <div className={offset}>
                        {body}
                    </div>
                </div>
            )
        } else {
            return body
        }
    }
}

/**
 * Display a button with a loading effect. Controlled by LoadingMultiState result
 */
export class ButtonStateFlip extends React.PureComponent {
    static defaultProps = {
        buttonClass: 'btn btn-primary',
        onClick: () => { },
        offset: 'col-md-offset-2 col-md-8',
        wrapGroup: false
    }
    static propTypes = {
        buttonClass: PropTypes.string,
        children: PropTypes.node,
        loadState: PropTypes.object.isRequired,
        offset: PropTypes.string,
        onClick: PropTypes.func,
        wrapGroup: PropTypes.bool
    }
    render() {
        var { buttonClass, children, loadState, onClick, offset, wrapGroup } = this.props

        let button
        if (loadState.state == 'loading') {
            button = <button className={buttonClass + ' m-progress'}>{children}</button>
        } else if (onClick) {
            button = <button className={buttonClass} onClick={onClick}>{children}</button>
        } else {
            button = <button className={buttonClass} type="submit">{children}</button>
        }

        if (wrapGroup) {
            return (
                <div className="form-group">
                    <div className={offset}>
                        {button}
                    </div>
                </div>
            )
        } else {
            return button
        }
    }
}

/**
 * MessagesComponent formats an array of messages for display.
 *
 * Use with error/warning messages
 */
export class MessagesComponent extends React.PureComponent {
    static propTypes = {
        messages: PropTypes.arrayOf(PropTypes.object),
        referenceString: PropTypes.string
    }

    render() {
        const {messages, referenceString} = this.props

        if (!messages) {
            return <div></div>
        }

        const errors = []
        const warnings = []

        messages.forEach((message, idx) => {
            let item = <li key={idx} className={message.level}>{message.message}</li>

            if (message.level == 'error') {
                errors.push(item)
            } else if (message.level == 'warning') {
                warnings.push(item)
            }
        })

        let errorString = ''
        if (errors.length > 0) {
            let title = referenceString ? `Errors ${referenceString}` : 'Errors'
            errorString = (
                <div>
                    <h5 className="error">{title}:</h5>
                    <ul>
                        {errors}
                    </ul>
                </div>
            )
        }

        let warningString = ''
        if (warnings.length > 0) {
            let title = referenceString ? `Warnings ${referenceString}` : 'Warnings'
            warningString = (
                <div>
                    <h5 className="warning">{title}:</h5>
                    <ul>
                        {warnings}
                    </ul>
                </div>
            )
        }

        return (
            <div>
                {errorString}
                {warningString}
            </div>
        )
    }
}

export class Alert extends React.PureComponent {
    static defaultProps = {
        alert: 'info',
        icon: 'info-sign'
    }
    render() {
        const { alert, icon, children } = this.props

        return (
            <div className={'alert alert-' + alert}>
                <span className={'glyphicon glyphicon-' + icon} aria-hidden="true"></span>
                {children}
            </div>
        )
    }
}

export class Glyphicon extends React.PureComponent {
    static propTypes = {
        icon: PropTypes.string
    }

    render() {
        return <span className={'glyphicon glyphicon-' + this.props.icon} aria-hidden="true"></span>
    }
}

export class Panel extends React.PureComponent {
    static defaultProps = {
        color: 'default',
        headingLevel: 'h3'
    }

    render() {
        const { color, heading, headingLevel, children } = this.props

        // doing an <h3 is simply syntactic sugar for react.createElement('h3', {propName:propVal, ...}, ...children)
        // So by directly using react.createElement we can create any heading from h1 to h6.
        const panelHeading = React.createElement(headingLevel, {className: 'panel-title'}, heading)

        return (
            <div className={'panel panel-' + color}>
                <div className="panel-heading">
                    {panelHeading}
                </div>
                <div className="panel-body">
                    {children}
                </div>
            </div>
        )
    }
}

export class Modal extends React.PureComponent {
    static propTypes = {
        title: PropTypes.string,
        footer: PropTypes.oneOfType([PropTypes.string, PropTypes.element, PropTypes.arrayOf(PropTypes.element)]),
        onClose: PropTypes.func
    }

    constructor(props) {
        super(props)
        this._body = window.$('body')
    }

    render() {
        const { title, children, footer, onClose } = this.props
        const footerContent = footer? <div className="modal-footer">{footer}</div> : null
        return (
            <div className="modal fade in" tabIndex="-1" role="dialog" style={{display: 'block'}}>
                <div className="modal-dialog" role="document">
                    <div className="modal-content">
                        <div className="modal-header">
                            <button type="button" className="close" aria-label="Close" onClick={onClose}>
                                <span aria-hidden="true">x</span>
                            </button>
                            <h4 className="modal-title">{title}</h4>
                        </div>
                        <div className="modal-body">
                            {children}
                        </div>
                        {footerContent}
                    </div>
                </div>
            </div>
        )
    }

    componentDidMount() {
        if (this._body) {
            this._body.addClass('modal-open')
            this._body.append('<div class="modal-backdrop fade in"></div>')
        }
    }

    componentWillUnmount() {
        if (this._body) {
            this._body.removeClass('modal-open')
            this._body.find('.modal-backdrop').remove()
        }
    }
}

export function scrollIntoView(id, after=null) {
    if (after) {
        setTimeout(() => scrollIntoView(id), after)
        return
    }
    if (window && window.document) {
        const elem = window.document.getElementById(id)
        if (elem && elem.scrollIntoView) {
            elem.scrollIntoView()
        }
    }
}
