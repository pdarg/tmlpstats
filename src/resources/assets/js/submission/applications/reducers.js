import { combineReducers } from 'redux'
import { modelReducer, formReducer } from 'react-redux-form'

import { appsCollection, applicationsLoad, saveAppLoad, messages } from './data'

const APPLICATIONS_BASE_KEY = 'submission.applications'
export const APPLICATIONS_FORM_KEY = APPLICATIONS_BASE_KEY + '.currentApp'


export const applicationReducer = combineReducers({
    loading: applicationsLoad.reducer(),
    applications: appsCollection.reducer(),
    currentApp: modelReducer(APPLICATIONS_FORM_KEY),
    forms: formReducer(APPLICATIONS_BASE_KEY),
    saveApp: saveAppLoad.reducer(),
    messages: messages.reducer(),
})
