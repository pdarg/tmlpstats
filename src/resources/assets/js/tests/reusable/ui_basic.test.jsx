import React from 'react'
import renderer from 'react-test-renderer'

import { ModeSelectButtons, LoadStateFlip } from '../../reusable/ui_basic'
import { LoadingMultiState } from '../../reusable/reducers'

test('ModeSelectButtons', () => {

    const items = [
        {id: 'foo', label: 'Hello'},
        {id: 'bar', label: 'Bar'}
    ]
    let onChange = () => null
    const tree = renderer.create(
        <ModeSelectButtons keyProp="id" items={items} current="foo" onClick={onChange} />
    ).toJSON()

    expect(tree).toMatchSnapshot()

    // Check that the 'active' class is only set on one of the buttons
    expect(tree.children.length).toEqual(2)
    expect(tree.children[0].props.className).toMatch(/active/)
    expect(tree.children[1].props.className).not.toMatch(/active/)
})


test('LoadStateFlip', () => {
    const states = LoadingMultiState.states

    const tree = renderer.create(
        <LoadStateFlip loadState={states.loading}>
            <div>Will never be seen</div>
        </LoadStateFlip>
    ).toJSON()
    expect(tree).toMatchSnapshot()
    expect(tree.children[1]).toMatch(/spinner/)

    const tree2 = renderer.create(
        <LoadStateFlip loadState={states.loaded}>
            <div className="foo">HELLO</div>
        </LoadStateFlip>
    ).toJSON()
    expect(tree).toMatchSnapshot()
    const div = tree2.children[0]
    expect(div.props.className).toEqual('foo')
    expect(div.children[0]).toEqual('HELLO')
})
