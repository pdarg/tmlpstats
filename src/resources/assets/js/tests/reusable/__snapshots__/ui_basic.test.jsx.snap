// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadStateFlip 1`] = `
<div>
  <span
    className="glyphicon glyphicon-send"
  />
  TODO spinner here...
</div>
`;

exports[`LoadStateFlip 2`] = `
<div>
  <span
    className="glyphicon glyphicon-send"
  />
  TODO spinner here...
</div>
`;

exports[`ModeSelectButtons 1`] = `
<div
  aria-label={undefined}
  className="btn-group"
  role="group"
>
  <button
    className="btn btn-default active"
    onClick={[Function]}
    type="button"
  >
    Hello
  </button>
  <button
    className="btn btn-default"
    onClick={[Function]}
    type="button"
  >
    Bar
  </button>
</div>
`;
